plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java

```diff
@@ -629,6 +634,41 @@ public ColumnMetadata getColumnMetadata(ConnectorSession session, ConnectorTable
                 .build();
     }
 
+    @Override
+    public void validateScan(ConnectorSession session, ConnectorTableHandle handle)
+    {
+        IcebergTableHandle table = (IcebergTableHandle) handle;
+        Schema schema = SchemaParser.fromJson(table.getTableSchemaJson());
+        Optional<PartitionSpec> partitionSpec = table.getPartitionSpecJson()
+                .map(partitionSpecJson -> PartitionSpecParser.fromJson(schema, partitionSpecJson));
+        if (partitionSpec.isEmpty() || partitionSpec.get().isUnpartitioned()) {
+            return;
+        }
+        if (isQueryPartitionFilterRequired(session) && table.getEnforcedPredicate().isAll()) {
+            Set<Integer> columnsWithPredicates = new HashSet<>();
+            table.getConstraintColumns().stream()
+                    .map(IcebergColumnHandle::getId)
+                    .forEach(columnsWithPredicates::add);
+            table.getUnenforcedPredicate().getDomains().orElse(Collections.emptyMap()).keySet().stream()
```

Comment:
Please note this change: checking the unenforced predicate seems to be necessary as well.
```
CREATE TABLE tbl (id integer, a varchar, b varchar, ds varchar) WITH (partitioning = ARRAY['bucket(ds, 16)'])
INSERT INTO tbl (id, a, ds) VALUES (1, 'a', 'a'), (2, 'b', 'b')
SELECT id FROM tbl WHERE ds = 'a'
``` 
In the above SELECT query, `ds` shows up as an unenforced predicate, and does not show up in the constraint columns.